gnomad/variant_qc/training.py

```diff
@@ -0,0 +1,130 @@
+
+    .. note::
+
+        - This function does not support multi-allelic variants.
+        - The function will give some stats about the TPs/FPs provided (Ti, Tv, indels).
+        - If a site qualifies as both TP and FP, it's excluded from both and therefore not in the returned Table.
+
+    :param ht: Input Table.
+    :param tp_expr: Expression for TP examples.
+    :param fp_expr: Expression for FP examples.
+    :param fp_to_tp: FP to TP ratio. If set to <= 0, all training examples are used.
+    :param test_expr: Optional expression to exclude a set of variants from training set. Still contains TP/FP label annotation.
+    :return: Table subset with corresponding TP and FP examples with desired FP to TP ratio.
+    """
+
+    ht = ht.select(
+        _tp=hl.or_else(tp_expr, False),
+        _fp=hl.or_else(fp_expr, False),
+        _exclude=False if test_expr is None else test_expr,
+    )
+    ht = ht.filter((ht._tp | ht._fp) & ~(ht._tp & ht._fp)).persist()
+
+    # Get stats about TP / FP sets
+    def _get_train_counts(ht: hl.Table) -> Tuple[int, int]:
+        """
+        Determine the number of TP and FP variants in the input Table and report some stats on Ti, Tv, indels.
+
+        :param ht: Input Table
+        :return: Counts of TP and FP variants in the table
+        """
+        train_stats = hl.struct(n=hl.agg.count())
+
+        if "alleles" in ht.row and ht.row.alleles.dtype == hl.tarray(hl.tstr):
+            train_stats = train_stats.annotate(
+                ti=hl.agg.count_where(
+                    hl.expr.is_transition(ht.alleles[0], ht.alleles[1])
+                ),
+                tv=hl.agg.count_where(
+                    hl.expr.is_transversion(ht.alleles[0], ht.alleles[1])
+                ),
+                indel=hl.agg.count_where(
+                    hl.expr.is_indel(ht.alleles[0], ht.alleles[1])
+                ),
+            )
+
+        # Sample training examples
+        pd_stats = (
+            ht.group_by(**{"contig": ht.locus.contig, "tp": ht._tp, "fp": ht._fp})
+            .aggregate(**train_stats)
+            .to_pandas()
+        )
+
+        logger.info(pformat(pd_stats))
+        pd_stats = pd_stats.fillna(False)
+
+        # Number of true positive and false positive variants to be sampled for the training set
+        n_tp = pd_stats[pd_stats["tp"] & ~pd_stats["fp"]]["n"].sum()
+        n_fp = pd_stats[~pd_stats["tp"] & pd_stats["fp"]]["n"].sum()
+
+        return n_tp, n_fp
+
+    n_tp, n_fp = _get_train_counts(ht.filter(~ht._exclude))
+
+    train_expr = True
+    if fp_to_tp > 0:
+        desired_fp = fp_to_tp * n_tp
+        if desired_fp < n_fp:
+            prob_tp = 1.0
+            prob_fp = desired_fp / n_fp
+        else:
+            prob_tp = n_fp / desired_fp
+            prob_fp = 1.0
+
+        logger.info(
+            f"Training examples sampling: tp={prob_tp}*{n_tp}, fp={prob_fp}*{n_fp}"
+        )
+
+        if prob_fp < 1.0:
```

Comment:
So I find the code below quite convoluted. I think the following does what we want and is much easier to read....
```
    test_expr = (
        hl.case(missing_false=True)
        .when(ht._fp & hl.or_else(~ht._tp, True), hl.rand_bool(prob_fp))
        .when(ht._tp & hl.or_else(~ht._fp, True), hl.rand_bool(prob_tp))
        .default(False)
    )
```
Note that `prob_tp = prob_fp = 1.0` will need to be added at line 90 too I think.

Reply:
Oh yes, thank you, that is so much easier to read